src/mesh/region_set.cc

```diff
@@ -0,0 +1,385 @@
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"New label (name) of the region. Has to be unique in one mesh.")
+		.declare_key("mesh_label", IT::String(), IT::Default::obligatory(),
+				"The mesh_label is e.g. physical volume name in GMSH format.")
+		.close();
+}
+
+
+
+const int RegionSetFromLabel::registrar =
+		Input::register_class< RegionSetFromLabel, const Input::Record &, Mesh * >("From_Label") +
+		RegionSetFromLabel::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetFromElements
+ */
+
+RegionSetFromElements::RegionSetFromElements(const Input::Record &rec, Mesh *mesh)
+: RegionSetBase(mesh)
+{
+	unsigned int region_id;
+	string region_label = rec.val<string>("name");
+	Region reg = mesh->region_db().find_label(region_label);
+
+	Input::Iterator<unsigned int> it = rec.find<unsigned int>("id");
+
+	if ( reg.is_valid() ) { // region with the specified label exists
+		region_id = reg.id();
+		if (it) {
+			if ( region_id != (*it) )
+				THROW( ExcInconsistentLabelId() << EI_Label(region_label) << EI_ID( *it ) << rec.ei_address()  );
+		}
+	} else {
+		if (it) {
+			region_id = (*it);
+		} else {
+			region_id = this->get_max_region_index();
+		}
+		stringstream ss;
+		ss << "declare at address \'" << rec.address_string() << "\'";
+		region_db_.add_region(region_id, region_label, RegionDB::undefined_dim, ss.str() );
+	}
+
+	Input::Array element_list;
+	if (rec.opt_val("element_list", element_list) ) {
+		std::vector<unsigned int> element_ids;
+		for (Input::Iterator<unsigned int> it_element = element_list.begin<unsigned int>();
+				it_element != element_list.end();
+		        ++it_element) {
+			std::map<unsigned int, unsigned int>::iterator it_map = el_to_reg_map_.find((*it_element));
+			if (it_map != el_to_reg_map_.end()) {
+				xprintf(Warn, "Region assigned to element with id %u will be rewritten.\n", (*it_element));
+			}
+			el_to_reg_map_.insert( std::make_pair((*it_element), region_id) );
+
+		}
+	}
+}
+
+
+
+const IT::Record & RegionSetFromElements::get_region_input_type()
+{
+    return IT::Record("From_Elements", "Region declared by name, ID and enum of elements.\n"
+    								   "Allows to get existing region or create new and assign\n"
+    								   "elements to its. Elements are specified by ids.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"Label (name) of the region. Has to be unique in one mesh.\n")
+		.declare_key("id", IT::Integer(0), IT::Default::optional(),
+				"The ID of the region to which you assign label.\n"
+				"If new region is created and ID is not set, unique ID will be generated automatically.")
+		.declare_key("element_list", IT::Array( IT::Integer(0) ), IT::Default::optional(),
```

Comment:
Any problem to make this key obligatory? Without the key it have no sense. Further the size of the list should be at least 1.